---
title: "ATAC-CutRun-ChIP 4"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_course_template/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: true    
---

```{r, setup,include=FALSE}
knitr::opts_chunk$set(message=F, warning=F) 
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```

## ATACseq Peak annotation

In these exercises we will review some of the functionality for genomic and functional annotation of peaks.

We will be using output from the differential peak analysis of ATACseq Week 6 vs Week 0. The starting table can be found in the 'data' folder of the course.  

<p>&nbsp;</p>
## Exercises


### 1. Peak annotation with ChIPseeker
  
Read in the differential output table 'ATAC_W6MinusD0.xlsx', make a GRanges and do the following:

  * map all of the consensus peak regions to genomic features (e.g. promoter, intron, exon, intergenic, etc.)
  * subset the peaks to those that increase in the W6 time point, and map these peaks to genomic features
  * make plots (or a single plot) to visualize these two distributions
    + Hint: by providing the plotAnnoBar function with a named list, multiple distributions can be visualized at once. 
  

```{r,echo=toMessage, results = 'hide', eval=T}
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)

W6MinusD0 <- rio::import("data/ATAC_W6MinusD0.xlsx")
W6MinusD0_gr <- GRanges(seqnames = W6MinusD0$seqnames,
                        IRanges(start = W6MinusD0$start, end = W6MinusD0$end),
                        log2FoldChange = W6MinusD0$log2FoldChange,
                        padj = W6MinusD0$padj)
peakAnno_all <- annotatePeak(W6MinusD0_gr, tssRegion=c(-1000, 1000), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")

W6MinusD0_gr_main <- W6MinusD0_gr[as.vector(seqnames(W6MinusD0_gr)) %in% paste0("chr", c(1:19, "X", "Y", "M"))]
W6MinusD0_gr_up <- W6MinusD0_gr_main[W6MinusD0_gr_main$log2FoldChange > 1 & W6MinusD0_gr_main$padj < 0.05, ]

peakAnno_up <- annotatePeak(W6MinusD0_gr_up, tssRegion=c(-1000, 1000), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")

plotAnnoBar(list(all_peaks = peakAnno_all, SigUp_peaks = peakAnno_up),
            title = "ATACseq peak distribution")
```

### 2. Functional enrichment of promoter peaks

Perform functional enrichment using the genes that have significant increases in ATACseq signal within peaks located in their promoters. 

  * Use the Biological Processes gene sets from the GO consortium as well as the Reactome gene sets available through the MsigDB R package. 
  * make a network plot to visualize each result
  * try out other gene sets that interest you from MSigDB. 


```{r instalssdflasaBSss2w,eval=T,echo=toMessage, results = 'hide'}
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(ggplot2)

# get universe of genes
allGeneGR <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
allGeneGR[1:2,]
allGeneIDs <- allGeneGR$gene_id


up_atac_GR <- as.GRanges(peakAnno_up)
up_atac_GR_prom <- up_atac_GR[grepl("Promoter", up_atac_GR$annotation)]
up_promGenes_uniq <- unique(up_atac_GR_prom$geneId)

GO_result_prom <- enrichGO(gene = up_promGenes_uniq, 
                           universe = allGeneIDs,
                           OrgDb = org.Mm.eg.db,
                           ont = "BP")

library(msigdbr)
msig_t2g_reac <- msigdbr(species = "Mus musculus", 
                    category = "C2", 
                    subcategory = "REACTOME")
msig_t2g_reac <- msig_t2g_reac[ , colnames(msig_t2g_reac) %in% c("gs_name", "entrez_gene")]

reactome_result_prom <- enricher(gene = up_promGenes_uniq, 
                                 universe = allGeneIDs,
                                 TERM2GENE = msig_t2g_reac)

```

GO BP output
```{r, eval=T,echo=toMessage}
data.frame(GO_result_prom, row.names = NULL)[1:3, 1:10]
GO_result_plot <- pairwise_termsim(GO_result_prom)
emapplot(GO_result_plot, showCategory = 20, cex.params = list(category_label = 0.6))

```

Reactome output:

  * no enriched terms!
```{r, eval=T,echo=toMessage}

data.frame(reactome_result_prom, row.names = NULL)[1:3, 1:10]
# no enriched terms!
```



### 3. Functional enrichment of all peaks with GREAT

  * using the bionomial test from GREAT, use the entire set of peaks that go up at W6 vs W0 and test enrichment for GO-Biological processes and Reactome gene sets.
  * make a bar plot showing the fold enrichment for the top 10 categories and fill the color of the bars based on the observed hits column.
  * try other gene sets from MSigDB.

```{r,echo=toMessage, results='hide', fig.width=8}

library(rGREAT)

great_gobp <- great(W6MinusD0_gr_up, 
                    gene_sets = "GO:BP", 
                    tss_source = "TxDb.Mmusculus.UCSC.mm10.knownGene")
great_gobp_tab <- getEnrichmentTables(great_gobp)
great_gobp_tab10 <- great_gobp_tab[1:10, ]
great_gobp_tab10$description <- ordered(great_gobp_tab10$description, levels = rev(great_gobp_tab10$description))


reac_gene_sets <- split(msig_t2g_reac$entrez_gene, msig_t2g_reac$gs_name)
reac_gene_sets <- lapply(reac_gene_sets, as.character)
great_reac <- great(W6MinusD0_gr_up, gene_sets = reac_gene_sets, tss_source = "TxDb.Mmusculus.UCSC.mm10.knownGene", mode = "oneClosest")

great_reac_tab <- getEnrichmentTable(great_reac)
great_reac_tab10 <- great_reac_tab[1:10, ]
great_reac_tab10 <- great_reac_tab10[order(great_reac_tab10$fold_enrichment, decreasing = TRUE), ]
great_reac_tab10$name_for_plot <- gsub("REACTOME_" ,"", great_reac_tab10$id)
great_reac_tab10$name_for_plot <- substr(great_reac_tab10$name_for_plot, 1, 50)
great_reac_tab10$name_for_plot <- ordered(great_reac_tab10$name_for_plot, levels = rev(great_reac_tab10$name_for_plot))

```


GO BP output:
```{r, eval=T,echo=toMessage, fig.width=8}
great_gobp_tab10[1:3, ]
ggplot(great_gobp_tab10, aes(x = description, y = fold_enrichment, fill = observed_region_hits)) + geom_col() + coord_flip()

```

Reactome output:
```{r, eval=T,echo=toMessage, fig.width=8}
great_reac_tab10[1:3, ]
ggplot(great_reac_tab10, aes(x = name_for_plot, y = fold_enrichment, fill = observed_region_hits)) + geom_col() + coord_flip()
```



