---
title: "My Exercise 1"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_course_template/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```


<p>&nbsp;</p>
These exercises are about [fastq quality control](https://rockefelleruniversity.github.io/ATAC.Cut-Run.ChIP/presentations/singlepage/Session1.html). 

**Exercise 1** - **Reading FastQs with ShortRead**

- Download fastq files for the second replicate of the Cut and Run from ENA. These are direct links for [Read1](ftp.sra.ebi.ac.uk/vol1/fastq/SRR201/017/SRR20110417/SRR20110417_1.fastq.gz) and [Read2](ftp.sra.ebi.ac.uk/vol1/fastq/SRR201/017/SRR20110417/SRR20110417_2.fastq.gz)).

- Read in a subset of Read 1 (lets just try 1000000 to keep things snappy).

```{r,echo=toMessage, eval=F}
library(ShortRead)

fqSample <- FastqSampler("~/Downloads/SRR20110417_1.fastq.gz",n=10^6)
fastq <- yield(fqSample)
fastq
```

```{r eval=T, echo=F}
#writeFastq(fastq,file = "data/SRR20110417_1_sub.fastq.gz",mode = "w")
fastq <- readFastq(dirPath = "data/SRR20110417_1_sub.fastq.gz")
fastq
```

- Lets review the first 10 reads. Lets look at ids, quality and sequence for the first ten reads.
```{r,echo=toMessage}
sread(fastq)[1:10]
quality(fastq)[1:10]
id(fastq)[1:10]
```
- Lets review the nucleotide frequency for each base in our read.

```{r,echo=toMessage}
library(ggplot2)
readSequences <- sread(fastq)
readSequences_AlpFreq <- alphabetFrequency(readSequences)
summed__AlpFreq  <- colSums(readSequences_AlpFreq)
summed__AlpFreq[c("A","C","G","T","N")]
readSequences_AlpbyCycle <- alphabetByCycle(readSequences)
AFreq <- readSequences_AlpbyCycle["A",]
CFreq <- readSequences_AlpbyCycle["C",]
GFreq <- readSequences_AlpbyCycle["G",]
TFreq <- readSequences_AlpbyCycle["T",]
toPlot <- data.frame(Count=c(AFreq,CFreq,GFreq,TFreq),
                     Cycle=rep(1:40,4),
                     Base=rep(c("A","C","G","T"),each=40))


ggplot(toPlot, aes(y=Count,x=Cycle,colour=Base)) + geom_line() +
  theme_bw()
```

- Compare this to read 2.



```{r,echo=toMessage, eval=F}
library(ShortRead)

fqSample <- FastqSampler("~/Downloads/SRR20110417_2.fastq.gz",n=10^6)
fastq2 <- yield(fqSample)
fastq2
```

```{r eval=T, echo=F}
#writeFastq(fastq,file = "data/SRR20110417_2_sub.fastq.gz",mode = "w")
fastq2 <- readFastq(dirPath = "data/SRR20110417_2_sub.fastq.gz")
fastq2

```

```{r,echo=toMessage}

readSequences <- sread(fastq2)
readSequences_AlpFreq <- alphabetFrequency(readSequences)
summed__AlpFreq  <- colSums(readSequences_AlpFreq)
summed__AlpFreq[c("A","C","G","T","N")]
readSequences_AlpbyCycle <- alphabetByCycle(readSequences)
AFreq <- readSequences_AlpbyCycle["A",]
CFreq <- readSequences_AlpbyCycle["C",]
GFreq <- readSequences_AlpbyCycle["G",]
TFreq <- readSequences_AlpbyCycle["T",]
toPlot <- data.frame(Count=c(AFreq,CFreq,GFreq,TFreq),
                     Cycle=rep(1:40,4),
                     Base=rep(c("A","C","G","T"),each=40))


ggplot(toPlot, aes(y=Count,x=Cycle,colour=Base)) + geom_line() +
  theme_bw()

```

**Exercise 2** - **Run Rfastp**

- Run Rfastp on our dataset with default settings.

```{r,echo=toMessage, eval=F}
library(Rfastp)

rfastp(read1="/Users/mattpaul/Downloads/SRR20110417_1.fastq.gz", read2 = "/Users/mattpaul/Downloads/SRR20110417_2.fastq.gz",outputFastq = "SOX9CNR_D0_rep2_filtered")
```

The html is [here]("data/SOX9CNR_D0_rep2_filtered.html")

- Try running Rfastp on our dataset with some extra parameters. Review the help page with ?rfastp. 

You could try:
* trimming from the front of the read
* adding in over representation analysis (this will take a little longer)

```{r,echo=toMessage, eval=F}

rfastp(read1="/Users/mattpaul/Downloads/SRR20110417_1.fastq.gz", read2 = "/Users/mattpaul/Downloads/SRR20110417_2.fastq.gz",outputFastq = "SOX9CNR_D0_rep2_filtered2",
       trimFrontRead1=10,
       trimFrontRead2=10,
       overrepresentationAnalysis=T)
```

The html is [here]("data/SOX9CNR_D0_rep2_filtered2.html")
