---
title: "CUT&RUN In Bioconductor (part5) <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_CUT&RUN/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no" 
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T) # delete cache before any merging 

```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# CUT&RUN/ATAC (part 5) - Peak annotation and functional enrichment

---
"    
  )
  
}

```


```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Peak Annotation

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Peak Annotation

---
"    
  )
  
}

```


## Annotation of peaks to genes

So far we have been working with CUT&RUN peaks corresponding to transcription factor binding or ATACseq peaks corresponding to open chromatin regions. Transcription factors, as implied in the name, can affect the expression of their target genes and open regions generally correlate with gene expression.

We will often annotate peaks to genes to try and identify the target of a transcription factor or a gene regulated by a regulatory element uncovered by ATACseq. This is typically done using a simple set of rules:

Peaks are typically annotated to a gene if
* They overlap the gene.
* The gene is the closest (and within a minimum distance).



---

## Peak annotation


A useful package for annotation of peaks to genes is **ChIPseeker**. 

By using pre-defined annotation in the from of a **TXDB** object for mouse (mm10 genome), ChIPseeker will provide us with an overview of where peaks land in the gene and distance to TSS sites.

First load the libraries we require for the next part and read in our SOX9 CUT&RUN peaks. 


```{r,echo=T, eval=T, echo=T, warning=FALSE,tidy=T,message=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(GenomeInfoDb)
library(ChIPseeker)
library(rtracklayer)

cnrPeaks_GR <- rtracklayer::import("data/peaks/SOX9CNR_W6_rep1_macs_peaks.narrowPeak")
```

---

## Peak annotation


The annotatePeak function accepts a GRanges object of the regions to annotate, a TXDB object for gene locations and a database object name to retrieve gene names from.


```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE}
peakAnno <- annotatePeak(cnrPeaks_GR, tssRegion=c(-1000, 1000), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")
class(peakAnno)
```
---
## Peak annotation

The result is a csAnno object containing annotation for peaks and overall annotation statistics.

```{r,eval=T,echo=T, message=F,messages=F, eval=T, echo=T, warning=FALSE,tidy=T}
peakAnno
```

---
## Peak annotation


The csAnno object contains the information on annotation of individual peaks to genes.

To extract this from the csAnno object the ChIPseeker functions *as.GRanges* or *as.data.frame* can be used to produce the respective object with peaks and their associated genes.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
annotatedPeaksGR <- as.GRanges(peakAnno)
annotatedPeaksDF <- as.data.frame(peakAnno)
```

---
## Peak annotation
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
annotatedPeaksGR[1:2,]
```

## Peak annotation

The genomic annotation for each peak is whin the *annotation* column and the closest gene is in shown in the *geneId*, *ENSEMBL*, and *SYMBOL* columns (geneId is the Entrez ID).

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
annotatedPeaksGR$annotation[1:5]
annotatedPeaksGR$SYMBOL[1:5]
```

---
## Vizualising peak annotation

Now we have the annotated peaks from ChIPseeker we can use some of ChIPseeker's plotting functions to display distribution of peaks in gene features. Here we use the **plotAnnoBar** function to plot this as a bar chart but  **plotAnnoPie** would produce a similar plot as a pie chart.


```{r, eval=T, echo=T, fig.height=5, fig.width=15, warning=FALSE, tidy=T}
plotAnnoBar(peakAnno)
```

---
## Vizualising peak annotation

Similarly we can plot the distribution of peaks around TSS sites.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,fig.height=5, fig.width=15,tidy=T}
plotDistToTSS(peakAnno)
```


---
# Vizualising peak annotation

ChIPseeker can also offer a succinct plot to describe the overlap between annotations.


```{r, eval=T, echo=T, fig.height=5, fig.width=15, warning=FALSE, tidy=T}
library(ggupset)
upsetplot(peakAnno, vennpie=F)
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gene Set Enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gene Set Enrichment

---
"    
  )
  
}

```

## Gene Set testing for peaks

Transcription factors or epigenetic marks may act on specific sets of genes grouped by a common biological feature (shared Biological function, common regulation in RNAseq experiment etc).

A frequent step in CUT&RUN or ATACseq analysis is to test whether common gene sets are enriched for transcription factor binding, epigenetic marks, or open chromatin regions.

Sources of well curated gene sets include [GO consortium](http://geneontology.org/) (gene's function, biological process and cellular localisation), [REACTOME](http://www.reactome.org/) (Biological Pathways) and [MsigDB](http://software.broadinstitute.org/gsea/msigdb/) (Computationally and Experimentally derived).

---
## Gene Set testing for peaks

Gene set enrichment testing may be performed on the sets of genes with peaks associated to them. We will not access these database libraries directly in testing but will use other R/Bioconductor libraries which make extensive use of them.

How we perform this analysis will depend on the type of peaks we are interested in. There are a wide range of types of peaks in our data set, some in promoters where annotation is straightforward, and many elswhere where annotation is trickier.


---
## Gene Set testing for peaks


We will go through two strategies:

 * using the closest gene with **ChIPseeker** followed by gene set enrichment with **clusterProfiler**. This is typically done for peaks in promoters
 
<div align="center">
<img src="imgs/gene_example_prom.png" alt="offset" height="250" width="800">
</div>


---
## Gene Set testing for peaks


We will go through two strategies:

 1. using the closest gene with **ChIPseeker** followed by gene set enrichment with **clusterProfiler**. This is typically done for peaks in promoters
 
 2. allowing for annotation of one peak with multiple genes using toolset called **GREAT**. This is usually done to annotate enhancer or distal peaks.
 
<div align="center">
<img src="imgs/gene_example_enh.png" alt="offset" height="250" width="800">
</div>

---
## Choosing a peak set

Our peak set is large (~72k peaks), which will result in many genes being annotated to peaks. 

Even from the ChIPseeker annotation of the nearest gene, there are almost 17k genes, making it unlikely we will find any real specific enrichment of gene sets.

We should choose a more specific set of peaks to test.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
length(unique(annotatedPeaksGR$geneId))
```
---


## Using differential peaks for test

The specific set of peaks we choose will depend on our question. In our case, we just performed differential analysis on our consensus peak set, so we can use the peaks that go up in W6 vs W0.

Here we read in the differential results and look at them.
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
W6MinusD0 <- rio::import("data/W6MinusD0.xlsx")

W6MinusD0[1:5, ]
```

---

## Using differential peaks for test

The gene annotation packages (e.g. ChIPseeker, GREAT) require a GRanges object. We can convert this table to a GRanges and keep key differential statistics as metadata in the GRanges object. 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
W6MinusD0_gr <- GRanges(seqnames = W6MinusD0$seqnames,
                        IRanges(start = W6MinusD0$start, end = W6MinusD0$end),
                        log2FoldChange = W6MinusD0$log2FoldChange,
                        padj = W6MinusD0$padj)

W6MinusD0_gr[1:3, ]
```

---

## Using differential peaks for test

We are going to look at the peaks that are increased in W6, so the GRanges is subset to the peaks that have a fold change greater than 2 and an adjusted p-value less than 0.05.

We also remove peaks that aren't on the main chromosomes (usually unplaced configs). 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
W6MinusD0_gr_main <- W6MinusD0_gr[as.vector(seqnames(W6MinusD0_gr)) %in% paste0("chr", c(1:19, "X", "Y", "M"))]
W6MinusD0_gr_up <- W6MinusD0_gr_main[W6MinusD0_gr_main$log2FoldChange > 1 & W6MinusD0_gr_main$padj < 0.05, ]

W6MinusD0_gr_up
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional Enrichment with nearest gene

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional Enrichment with nearest gene

---
"    
  )
  
}

```

## Enrichment with nearest gene

To perform gene set using the closest gene to each peak, we will use the **clusterProfiler package**.

clusterProfiler provides multiple enrichment functions that allow for comparison of your gene list to known (e.g. GO, KEGG) or custom gene sets. Detailed information about all of the functionality within this package is available [here](http://yulab-smu.top/clusterProfiler-book/).

The functions in clusterProfiler take a list of genes, which we can obtain from the ChIPseeker result. 

---
## Identify universe of genes

Pathway enrichment requires a 'universe' of genes. We can extract all genes which are included in the TxDb object to use as our universe.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T, message = F}
allGeneGR <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
allGeneGR[1:2,]
allGeneIDs <- allGeneGR$gene_id
```

---

## Enrichment with nearest gene

To get a list of potential target genes, we again use the **ChIPseeker package** to associate our peaks, representing potential transcription factor binding events, to their overlapping or closest mm10 genes.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE, message = F, results='hide'}
peakAnno_up <- annotatePeak(W6MinusD0_gr_up, 
                            tssRegion=c(-1000, 1000),
                            TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                            annoDb="org.Mm.eg.db")

```

---

## Enrichment with nearest gene

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE, message = F}

peakAnno_up
```

---

## Enrichment with nearest gene

We again convert this csAnno object to a GRanges to see the individual peak annotations.

```{r}
annotatedPeaksGR_up <- as.GRanges(peakAnno_up)
annotatedPeaksGR_up[1:3]
```

---
## Analysis of promoter peaks 

The closest gene annotation approach works well when we subset to just the peaks localized to promoters as this leads to confident peak to gene annotations.  

For this Sox9 CUT&RUN dataset there are not many promoter-bound peaks, resulting in a small gene list. This might be okay if certain gene sets are enriched in these genes.
```{r}
annotatedPeaksGR_up_prom <- annotatedPeaksGR_up[grepl("Promoter", annotatedPeaksGR_up$annotation)]
up_promGenes_uniq <- unique(annotatedPeaksGR_up_prom$geneId)

length(up_promGenes_uniq)
```


---
## Analysis of promoter peaks 

Once we have our gene list and the universe of genes in the same format, we can use them in the `enrichGO` function to perform gene ontology analysis

For the **ont** argument, we can choose between the "BP", "MF", and "CC" subontologies, or "ALL" for all three.

```{r, message = F, warning=F, results = 'hide'}
library(clusterProfiler)
library(org.Mm.eg.db)

GO_result_prom <- enrichGO(gene = up_promGenes_uniq, 
                      universe = allGeneIDs,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP")

```

---
## Analysis of promoter peaks 

In our case, this results in no enriched terms. This suggests the Sox9 is not likely not enacting its main regulatory effect by binding to promoters.

This is not surprising given how few peaks overlapped with promoters 

```{r, message = F, warning=F}

GO_result_prom
```

---
## Nearest gene approach with all genes

We can try to use the whole set of Sox9 peaks that go up in the week 6 time point. 

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}

up_genes_uniq <- unique(annotatedPeaksGR_up$geneId)

length(up_genes_uniq)
```

---
## Nearest gene approach with all genes

We again use the `enrichGO` function, but this time with all of the genes from the differential peaks. 

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}

GO_result <- enrichGO(gene = up_genes_uniq, 
                      universe = allGeneIDs,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP")
GO_result
```

---
## Nearest gene approach with all genes

We now have an **enrichResult** instance. From this object, we can extract a data frame of the most highly enriched gene ontology categories.
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
GO_result_df <- data.frame(GO_result, row.names = NULL)
GO_result_df[1:3, ]
```

---
## Visualize enrichment result

Network plots can be generated from any enrichResult object using the **enrichplot** package.

We measure similarities between the various significant gene sets and group them accordingly. The **showCategory** argument specifies how many top gene ontology hits to show. 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T, fig.height=4, fig.width=9}
library(enrichplot)
GO_result_plot <- pairwise_termsim(GO_result)
emapplot(GO_result_plot, showCategory = 20, cex_label_category = 0.6)
```


---
## Using MSigDB gene sets

In addition to gene ontology, we can test our gene list against custom gene sets that we import using the clusterProfiler `enricher` function. Similar to the enrichGO function, this will produce an enrichResult object that can be used for visualization. 

Here we will get gene sets from MSigDB using the **msigdbr** package. We can use this package to pull out specific gene sets, which can be seen [at this link](http://software.broadinstitute.org/gsea/msigdb/). We can also run the `msigdbr_collections` function to see the categories and subcategory codes that will be used for accessing the gene sets. 

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(msigdbr)
msigdbr_collections()
```

---

## Using MSigDB gene sets

From the data frame on the previous slide we can identify the category/subcategory we want, and use those in the `msigdbr` function. Here we will use "C2" as the category, and "REACTOME" as the subcategory to access the [Reactome](https://reactome.org/) gene sets, and in the end we need to get a data frame where the first column contains the name of the gene sets and the second column contains the gene ID. 

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(msigdbr)
msig_t2g_reac <- msigdbr(species = "Mus musculus", 
                    category = "C2", 
                    subcategory = "REACTOME")
msig_t2g_reac <- msig_t2g_reac[ , colnames(msig_t2g_reac) %in% c("gs_name", "entrez_gene")]
msig_t2g_reac[1:3, ]
```

---
## Using MSigDB gene sets

We then run the gene set enrichment, using the term to gene mapping we created as the **TERM2GENE** argument in the enricher function. 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
reactome <- enricher(gene = up_genes_uniq, 
                     universe = allGeneIDs,
                     TERM2GENE = msig_t2g_reac)
reactome_df <- data.frame(reactome, row.names = NULL)
# clean up table to print on slide
reactome_df <- reactome_df %>% dplyr::select(-Description, -geneID) %>% dplyr::mutate(ID = substr(ID, 1, 50))
reactome_df[1:5, ]

```
---

## Limitations of nearest gene

In our case, annotating the peaks with the nearest gene is likely not the best strategy given the high number of non-promoter and distal intergenic peaks. 

For example, it's known that regulatory elements like enhancers can regulate genes that are far away, and can even skip over nearby genes to regulate distant ones. 

By only using the closest gene, we might be missing key peak to gene relationships. For enhancer-focused peak sets we have other methods that are better suited. 




---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional Enrichment of distal regions with GREAT

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional Enrichment of distal regions with GREAT

---
"    
  )
  
}

```

## GREAT for GO and functional testing

For methods that rely on genomic regions (i.e. peaks) like CUT&RUN and ATACseq, tools have been developed to better incorporate peaks distal to genes into their enrichment testing, such as the popular [GREAT](http://bejerano.stanford.edu/great/public/html/splash.php) toolset.

Incorporating distal peaks by rules such as nearest gene results in some genes having a higher chance of being selected and hence some gene sets as a whole having a higher chance of having its members selected.

It's known that genes can be regulated by elements that are far away (e.g. 1000kb), so you might miss key regulatory relationships by limiting to the closest gene. 

Of course this method will likely introduce more false positives, so this might be worth it if you know you are looking at enhancers (like Sox9), but not if the TF is binding to promoters. 

---


## GREAT for GO and functional testing

Instead of assigning a peak the closest genes (like ChIPseeker), GREAT provides flexible options for controlling peak annotation.

The default setting is called 'basalPlusExt', which gives each gene a defined basal regulatory regions (e.g. 5kb upstream and 1kb downstream) and then expends that region in either direction based on a maximum set distance (e.g. 1000kb).

A peak is then annotated by any gene regulatory region the peak overlaps. This might be multiple genes.

<img src="imgs/great_basal_extension.png" alt="offset" height="250" width="800">

---

## GREAT for GO and functional testing

After defining regulatory regions for each individual gene, GREAT then compares the proportion of peaks mapping to an entire gene set's regulatory regions to the proportion of the genome occupied by gene set's regulatory regions.

i.e. If a gene set's regulatory regions account for 1% of the genome then one might expect 1% of peaks to overlap these regions by chance.

This will help reduce bias towards gene sets that have many genes with large regulatory domains (i.e. more distal, few surrounding genes)
---


## rGREAT 


We can use the rGREAT package to utilize the GREAT algorithm. 

There are two ways to use rGREAT, by remotely accessing the GREAT server with the `submitGreatJob` function, or running the algorithms locally with the `great` function. 

The local version provides a bit more flexibility, so we will use this. 

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
library(rGREAT)
```


---

## rGREAT for GO and functional testing


To use the `great` function we need at least three arguments: 

  * GRanges of Sox9 peaks
  * the gene sets to test against (look at the help to see options - ?great)
  * a genome to define the TSS loci, usually easiest to specify a TxDb 

This function contains many other arguments to fine tune the gene annotation settings.

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
great_gobp <- great(W6MinusD0_gr_up, gene_sets = "GO:BP", tss_source = "TxDb.Mmusculus.UCSC.mm10.knownGene")

```

---

## rGREAT for GO and functional testing

This function returns a GreatObject containing the results of the test and various informaiton about the inputs. These can be accessed using the '@' accessor. 

If we print the object, we get a summary.

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
great_gobp
```

---

## rGREAT for GO and functional testing


The results table can be retrieved using the `getEnrichmentTables` function.


```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
great_gobp_tab <- getEnrichmentTables(great_gobp)

great_gobp_tab[1:5, ]

```


---

## Reactome gene sets with rGREAT 

The `great` function can be used to test custom gene sets. The gene sets need to be a list where each entry is named and contains a character vector of gene IDs. 

We can convert our Reactome gene set table we previously generated to a list. 

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
# convert to a list of gene sets
reac_gene_sets <- split(msig_t2g_reac$entrez_gene, msig_t2g_reac$gs_name)
reac_gene_sets <- lapply(reac_gene_sets, as.character)  # just to make sure gene IDs are all in character.

reac_gene_sets[1:2]
```

---


## Reactome gene sets with rGREAT 

This list is then used for the 'gene_sets' argument for the `great` function.

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}

great_reac <- great(W6MinusD0_gr_up, gene_sets = reac_gene_sets, tss_source = "TxDb.Mmusculus.UCSC.mm10.knownGene", mode = "oneClosest")

great_reac_tab <- getEnrichmentTable(great_reac)
great_reac_tab[1:2, ]
```


---

## Reactome gene sets with rGREAT 

Our top hit, RUNX related regulation, lines up well with the publication. Below is a figure showing RNAseq results, demonstrating that at week 6 there is a string increase in RUNX genes. 


<img src="imgs/rnaseq_heatmap.png" alt="igv" height="500" width="300">


---


## Retrieve rGREAT gene annotations


It might be useful to know the genes associated with each peak. For example to overlap with a DE gene list from an RNAseq experiment. 

This can be retrieved with the `getRegionGeneAssociations` function. 


```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
great_genes <- getRegionGeneAssociations(great_reac)

great_genes[1:5]
```


---

## Time for an exercise!

Exercise on CUT&RUN data can be found [here](../../exercises/exercises/CUT&RUN_part1_exercise.html)

---
## Answers to exercise

Answers can be found [here](../../exercises/answers/CUT&RUN_part1_answers.html)

R code for solutions can be found [here](../../exercises/answers/CUT&RUN_part1_answers.R)


