---
title: "Epigenomics, Session 6"
subtitle: " <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html> Bioinformatics Resource Center - Rockefeller University"
author: 
  - "http://rockefelleruniversity.github.io/RU_course_template/"
  - "brc@rockefeller.edu"
output: 
  xaringan::moon_reader:
    css: ["default.css", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      titleSlideClass: [middle, inverse]
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# Epigenomics (part 6)

---
"    
  )
  
}

```


##  Recap

?? can add a recap to link to keypoints in first session??
Session 1 covered introduction to R data types and import/export of data.

- [Background to R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#background-to-r)
- [Data types in R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#data_types_in_r)
- [Reading and writing in R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#reading-and-writing-data-in-r)

---
##  Recap

1) Fastq QC
2) Alignment
3) Peak calling
4) Technique QC
5) Consensus Building
6) Counting
7) Differentials
8) Annotation and enrichment

---
## Our Data

We have been working to process and a characterize developmental changes in the context of the TF Sox9 using data from the Fuchs lab: [*The pioneer factor SOX9 competes for epigenetic factors to switch stem cell fates*](https://www.nature.com/articles/s41556-023-01184-y)


---
## Motifs

Once we have identified regions of interest from our ATAC or Cut&Run often the next step is to investigate the motifs enriched under peaks. 
Motif analysis like this can help find the drivers of epigenomic changes and help create a more mechanistic understanding of your experiment.

For Cut&Run with a known transcription factor this kind of analysis may be less obvious as we have an expected target i.e. Sox9. That said it is still useful to validate our IP, find cofactors, indirect effects and also find specific motif variants.


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Motif Databases

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Motif Databases

---
"    
  )
  
}
```

## Known Motif sources

Bioconductor provides two major sources of motifs as database packages.

These include:

 * The [MotifDb](https://www.bioconductor.org/packages/release/bioc/html/MotifDb.html) package.
 * The JASPAR databases, [JASPAR2024](https://www.bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html) being latest (they do biannual updates).

---
## MotifDb

The MotifDB package collects motif information from a wide range of sources and stores them in a DB object for use with other Bioconductor packages.

```{r,eval=TRUE}
library(MotifDb)
MotifDb
```

---
## MotifDb

MotifDb object is special class of object called a **MotifList**.


```{r,eval=TRUE}
class(MotifDb)
```


---
## MotifDb

Like standard List objects we can use length and names to get some information on our object


```{r,eval=TRUE}
length(MotifDb)
MotifNames <- names(MotifDb)
MotifNames[1:10]
```

---
## Accesing MotifDb contents

We can also access information directly from our list using standard list accessors.

Here a **[** will subset to a single MotifList. Now we can see the information held in the MotifList a little more clearly.

```{r,eval=TRUE}
MotifDb[1]
```



---
## Accesing MotifDb contents

A **[[** will subset to object within the element as with standard lists. Here we extract the position probability matrix.


```{r,eval=TRUE}
MotifDb[[1]]
colSums(MotifDb[[1]])
```

---
## Accesing MotifDb contents

We can extract a DataFrame of all the motif metadata information using the **values()** function.


```{r,eval=TRUE}
values(MotifDb)[1:2,]
```

---
## Accesing MotifDb contents

We can use the **query** function to subset our MotifList by infomation in the metadata.


```{r,eval=TRUE}
Sox9Motifs <- query(MotifDb,"Sox9")
Sox9Motifs
```

---
## Accesing MotifDb contents

For more specific queries, multiple words can be used for filtering.


```{r,eval=TRUE}
Sox9Motifs <- query(MotifDb,c("Sox9","hsapiens","jaspar2022"))
Sox9Motifs
```


---
## JASPAR2020 package

The JASPAR packages are updated more frequently and so may contain motifs not characterized in the MotifDb package.

We can load the **JASPAR** package as usual. This is a little more complex then MotifDb, but ultimately we are connecting to a SQL database. 

```{r,eval=TRUE}
library(JASPAR2024)
library(RSQLite)

JASPAR2024 <- JASPAR2024()
sq24 <- RSQLite::dbConnect(RSQLite::SQLite(), db(JASPAR2024))


```

---
## JASPAR2020 and TFBSTools

To interact with JASPAR package we will make use of the **TFBSTools** package from the same lab.

Whereas the JASPAR package holds the information on Motifs and Position Probability Matrices (PPMs), TFBSTools has the functionality to manipulate and interact with these tools.

Three useful functions available from TFBStools to interact with the JASPAR databases are the **getMatrixSet**, **getMatrixByID** and **getMatrixByID**.


```{r,eval=TRUE, message=F, warning=F}
library(TFBSTools)
?getMatrixByID
```


---
## TFBSTools for TFs

The **getMatrixByID** and **getMatrixByName** take the JASPAR DB object and a JASPAR ID or transcription factor name respectively. If you are unsure of exact names/IDs you can review the Jaspar [website](https://jaspar.elixir.no/matrix/MA0077.1/) first.

Here we are using the transcription factor SOX9 again. The result is a new object class **PFMatrix**.

```{r,eval=TRUE,echo=TRUE}
SOX9mat <- getMatrixByName(sq24,"SOX9")
class(SOX9mat)

```

---
## TFBSTools for motifs

JASPAR IDs are unique, so we can use them to select the exact motif we wanted. Sox9 has two very subtlely different motifs

```{r,eval=TRUE,echo=TRUE}
SOX9mat2 <- getMatrixByID(sq24,"MA0077.1")
```

---
## TFBSTools for motifs

List accessors will not work here but we can retrieve names using the **ID** function.

```{r,eval=TRUE,echo=TRUE}
ID(SOX9mat)
```


---
## Position Frequency Matrix

To get hold of the Position Frequency Matrix (PFM) we can use the **Matrix** or **as.matrix** functions.

```{r,eval=TRUE}
myMatrix <- Matrix(SOX9mat)
myMatrixToo <- as.matrix(myMatrix)
myMatrix
```

---
## TFBSTools for motif sets

We can also use the **getMatrixSet()** function to retrieve sets of motifs. We can specify a list of options for motifs we want to retrieve. 

To see the available filters use the help for getMatrixSet() function, ?getMatrixSet.

Here we retrieve the vertebrate, JASPAR CORE motifs. We additional specify *all_versions* is FALSE to only include the latest version of a motif.

```{r,eval=TRUE}
opts <- list()
opts[["collection"]] <- "CORE"
opts[["tax_group"]] <- "vertebrates"
opts[["all_versions"]] <- FALSE
motifList <- getMatrixSet(sq24, opts)
   
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Visualizing motifs

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Visualizing motifs

---
"    
  )
  
}
```

## Visualizing motifs

The seqLogo package offers a simple and intuitive way of visualizing our base frequencies within out motifs using the **seqLogo** function.

A simple seqLogo shows the relative frequency of a base at each motif position by the relative size of the base compared to other bases at the same position. 

```{r,eval=TRUE,echo=FALSE,warning=FALSE,message=FALSE, fig.height=4,fig.width=6}
require(seqLogo)
Sox9Motifs <- query(MotifDb,"Sox9")
seqLogo::seqLogo(Sox9Motifs[[1]],ic.scale = FALSE)
```

---
## Visualizing motifs

To create a seqLogo we must supply a PPM to the **seqLogo** function. Here we grab a PPM straight from MotifdDb for Sox9. 

Here we set the **ic.scale** to FALSE to show the probability across the Y-axis

```{r,eval=TRUE, fig.height=4,fig.width=6}
library(seqLogo)
Sox9Motifs <- query(MotifDb,"Sox9")
seqLogo::seqLogo(Sox9Motifs[[1]], ic.scale = FALSE)
```

---
## Visualizing motifs

It may be useful to plot the probabilities as information content. Information content here will range from 0 to 2 Bits. Positions with equal probabilities for each base will score 0 and positions with only 1 possible base will score 2. 

This allows you to quickly identify the important bases.

```{r,eval=TRUE, fig.height=4,fig.width=6}
seqLogo::seqLogo(Sox9Motifs[[1]])
```

---

## Visualizing motifs

The TFBSTools package and JASPAR provide us with point frequency matrices which we can't use directly in the seqLogo package

```{r,eval=TRUE}
myMatrix
```

---
## Visualizing motifs

We can convert our point frequency matrix to a point probabilty matrix by simply dividing columns by their sum.

```{r,eval=TRUE,echo=TRUE, fig.height=4,fig.width=6}
ppm <- myMatrix/colSums(myMatrix)
ppm
```

---

## Visualizing motifs

We can then plot the result matrix using seqLogo.

```{r,eval=TRUE,echo=TRUE, fig.height=4,fig.width=6}
seqLogo::seqLogo(ppm)
```

---
## Visualizing motifs

Fortunately, TFBSTools has its own version of seqLogo function we can use with one of its own **ICMatrix** classes. We simply need to convert our object with the **toICM()** function.

```{r,eval=TRUE,echo=TRUE, fig.height=4,fig.width=6}
Sox9_IC <- toICM(SOX9mat)
TFBSTools::seqLogo(Sox9_IC)
```

---

## Visualizing with ggplot

We can also use a ggplot2 style grammar of graphics with the **ggseqlogo** package.

```{r,eval=TRUE,echo=TRUE, fig.height=4,fig.width=6}
library(ggseqlogo)
library(ggplot2)
ggseqlogo(myMatrix)+theme_minimal()
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Motif enrichment analysis

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Motif enrichment analysis

---
"    
  )
  
}
```


## Motif review

Motif enrichment can be performed using the [MEME suite](https://meme-suite.org/meme/) of tools. They have developed a huge collection of algorithms with situation specificity for finding and matching motifs or detecting differential enrichment of motifs. 

You can directly run these tools sing the web interface (and we show you how to work with the results [here](https://rockefelleruniversity.github.io/RU_ATACseq/presentations/singlepage/RU_ATAC_part3.html)), but in these sessions we will show you how to run this through R. 

---
## Installing MEME

As we did for MACS, we can also install MEME from the Conda repository using the Herper package.

```{r, eval=F}
library(Herper)
install_CondaTools("meme","motif_analysis", channels="bioconda", pathToMiniConda = "miniconda")
```



---
## Motif DBs for MEME

Now we have MEME installed, the next step is to grab some Motifs as our reference. We can follow similar steps to before. 

```{r}

opts <- list()
opts[["tax_group"]] <- "vertebrates"
opts[["collection"]] <- "CORE"
opts[["all_versions"]] <- FALSE

jaspar <- JASPAR2024()
sq24 <- RSQLite::dbConnect(RSQLite::SQLite(), db(jaspar))

```

---
## Motif DBs for MEME

We specifically will want to convert this to a new format: the *universalmotif* class. We can do this by grabbing our motifs like we did for TFBStools, then using the *convert_motifs* function. The result is a list of universalmotif PCMs. 

```{r}

library(universalmotif)
motifs <- getMatrixSet(sq24, opts)
motifs_um <- convert_motifs(motifs)
class(motifs_um)
class(motifs_um[[1]])
```

---
## Query Sequences for MEME

We need a collection of sequences that we are querying. In this case our query is: what motifs are enriched in our regulated regions at W6 time point. We can start by importing the ranges from the data folder *UpinW6.bed*.

```{r}

library(rtracklayer)

UpinW6_peaks <- rtracklayer::import("data/UpinW6.bed")

```

---
## Query Sequences for MEME

Next we need to grab the corresponding sequence from under these ranges. Where you are grabbing sequences from needs to match your alignment reference. If you have updated your BSGenome or are using a different build you can grab the sequence ifnromation from any fasta using *readDNAStringSet*

```{r, eval=F}
library(Biostrings)
library(BSgenome.Mmusculus.UCSC.mm10)

peaksSequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, UpinW6_peaks)
names(peaksSequences) <- paste0("peak_", 1:length(UpinW6_peaks))
```


```{r, eval=F, echo=F, warning=F, message=F}
library(Biostrings)
mm10 <- readDNAStringSet("~/Downloads/mm10.fa.gz")

peaksSequences <- getSeq(mm10, UpinW6_peaks)
names(peaksSequences) <- paste0("peak_", 1:length(UpinW6_peaks))
```

---
## Background Sequences for MEME

The last thing we need is background sequences. Though you can use a random shuffled background it is often better to give something a little more representative of your data i.e. consensus peaks. Otherwise common motifs found in regions like promoters may seem enriched, even if they are not enriched in the specific promoters you care about.

```{r, eval=F}
HC_peaks <- rtracklayer::import("data/HC_Peaks.bed")
background_peaksSequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, HC_peaks)
names(background_peaksSequences) <- paste0("peak_", 1:length(HC_peaks))
```


```{r, eval=F, echo=F, warning=F, message=F}
HC_peaks <- rtracklayer::import("data/HC_Peaks.bed")
background_peaksSequences <- getSeq(mm10, HC_peaks)
names(background_peaksSequences) <- paste0("peak_", 1:length(HC_peaks))
```


---
## Running MEME

Though there is no R pacakge for MEME, there is a wrapper package that will run MEME for you from within R. This is called *memes*. We just need to be able to tell memes where MEME is. 

```{r, eval=F}
library(memes)

meme_path_id="miniconda/envs/motif_analysis/bin"
```

---
## Running MEME

We will use the [**Ame** algorithm](https://meme-suite.org/meme/tools/ame) (Analysis of Motif Enrichment) to look for enriched motifs.

*NOTE: This will take some time. We will provide the result*

```{r, eval=F, echo=FALSE}

meme_path_id ="/Users/mattpaul/Documents/RU/Software/mini/envs/motif_analysis/bin"

```


```{R, eval=F}
ame_results <- runAme(peaksSequences,
                          control = background_peaksSequences,
                          outdir = "ame_diff",
                          meme_path = meme_path_id,
                          database = motifs_um)
```

```{r, eval=F, echo=F}

save(ame_results, file = "data/ame_results.RData")

```


---
## Reviewing AME results

Once this has run the MEME results are saved in the output directory we specified in the function. *data/ame_diff*. We can review the [report]("data/ame_diff/ame.html") to see our top hits.

---
## Reviewing AME results

The memes package also returns our results into R. We have saved the R result object *data/ame_results.RData* so you can load it in. 

This object is a special kind of data frame called a tibble. This means we can use a combination of tidyverse and memes functions to extract our result and generate some nice plots.

```{r}

load(file = "data/ame_results.RData")

class(ame_results)

```

---
## Reviewing AME results

First lets look at our top 10 motifs associated with the W6 timepoint. They are predominately the SOX family, which is a good result. 

```{r}

ame_results %>% 
  dplyr::filter(rank %in% 1:10) %>% 
  plot_ame_heatmap()  +
    ggtitle("Top 10 AME Hits")
```

---
## Reviewing AME results

We can then use what we know about making motifs to quickly grab our seqlogo.

```{r}

top_res_motif <- getMatrixByID(sq24,ame_results$motif_alt_id[1])
top_res_motif_mat <- Matrix(top_res_motif)
ggseqlogo(top_res_motif_mat)+theme_minimal()

```

---
## A quick note

As you can see we have a high degree of redundancy in our results i.e. all the top hits are different Sox!

This can make it a little trickier to aprse your results. Two factors can help with this.

1) Filter out unexpressed genes. It is possible to use matching RNAseq (or even ATAC) to filter out TF motifs if they are not expressed. This will also boost your padj scores. 

2) There are [JASPAR familial profiles](https://jaspar.elixir.no/matrix-clusters/) that contain a consensus motif for TF families. This will simplify the result, though it may blunt and obfuscate some findings. 

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# De novo motif analysis

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# De novo motif analysis

---
"    
  )
  
}
```


## De novo motifs

Databases of motifs are a great resource, but sometimes they fall short or miss aditional biology for various reasons. 

* Uncharacterized motif
* Rare motif variants
* Motif evolution versus model organism
* Cofactor binding

By doing de novo motif characterization you can account for this. The [*DREME* algorithm](https://meme-suite.org/meme/tools/dreme) from MEME is used for this.

---
## Running DREME

We can run DREME (Discriminative Regular Expression Motif Elicitation) in a very similar way to AME, using the *runDreme()* function. The only change is we do not include a motif reference database.  


```{R, eval=F}
dreme_results <- runDreme(peaksSequences,
                          control = background_peaksSequences,
                          outdir = "dreme_diff",
                          meme_path = meme_path_id)
```

```{r, eval=F, echo=F}

save(dreme_results, file = "data/dreme_results.RData")

```

---
## Reviewing Dreme results

As before, once this has run the MEME results are saved in the output directory we specified in the function. *data/dreme_diff*. We can review the [report]("data/dreme_diff/dreme.html") to see our top hits.

---
## Reviewing DREME results

The memes package also returns our results into R. We have saved the R result object *data/dreme_results.RData* so you can load it in. 

This object again is a special kind of data frame, but this time it is a universalmotif data frame as it has motifs stored inside.

```{r}

load(file = "data/dreme_results.RData")

class(dreme_results)

```

---
## Reviewing DREME results

DREME results will be sightly different to AME as we do not know the identity of the motif. But we can view the motifs easily using some universalmotif functions.

```{r}
dreme_results[1:3,] %>% 
  to_list() %>% 
  view_motifs()
```

---
## Matching DREME motifs

Once you have derived a list of de novo motifs, the next step is then to compare them to a list of known motifs. We can use the [*TomTom*](https://meme-suite.org/meme/tools/tomtom) algorithm from MEME to do this.  

```{r}

tomtom_results <- runTomTom(dreme_results,
            meme_path = meme_path_id,
            database = motifs_um,
            outdir = "tomtom_from_dreme"
            )

```

---
## Reviewing TOMTOM results

The *best_match_motif* column contains all the top hits from TOMTOM for each de novo motif. 

```{r}
view_motifs(tomtom_results$best_match_motif[1:3])

```

---
## Reviewing TOMTOM results

Though there is a top hit, each motif can have multiple hits. These are some of the many potential hits:

```{r}
unlist(lapply(tomtom_results$tomtom[[1]]$match_motif, function(x) x@name))

```

---
## Reviewing TOMTOM results

If you wanted to review these motifs you can use the *view_tomtom_hits* function. This allows you to compare the motifs easily to the original. Though our result is pretty clear, this is an important step as the "best match" is not always the right match. You need to make sure it amkes sense given your domain knowledge.

```{r, eval=F}

tomtom_results[1,] %>% 
  view_tomtom_hits(top_n = 3)
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Finding Motifs

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Finding Motifs

---
"    
  )
  
}
```


## Motif Locations

We now have a couple of tools in our toolkit to characterize which motifs are interesting to us. Often the next step is to actually determine where those motifs are. 

Unsurprisingly, MEME also has a tool for this: [FIMO](https://meme-suite.org/meme/tools/fimo). You can use the [*runFIMO()*](https://www.bioconductor.org/packages/devel/bioc/vignettes/memes/inst/doc/core_fimo.html) function to do this through memes. 

There is also a nice alternative tool called motifmatchr, which we will show you as it is very quick. 

---
## motifmatchr

To identify known motifs we will use the **motifmatchr** package which is a wrapper for the MOODS c++ library.

This means that motifmatchr offers us a fast methods to identify motifs within our data.

We will use motifmatchr/MOODs with a default p-value cut-off.

---
## motifmatchr

Again lets use the JASPAR database as our reference for this. For motifmatchr we want a PFMatrixList using TFBSTools.

```{r}
opts <- list()
opts[["tax_group"]] <- "vertebrates"
opts[["collection"]] <- "CORE"
opts[["all_versions"]] <- FALSE

jaspar <- JASPAR2024()
sq24 <- RSQLite::dbConnect(RSQLite::SQLite(), db(jaspar))
motifs <- getMatrixSet(sq24, opts)
```

---
## motifmatchr

The motifmatchr package main function is **matchMotifs()**.

As with many Bioconductor functions , **matchMotifs** makes use of other Bioconductor objects such the BSGenome, GRanges and summarizedExperiment objects.

We can review the full functionality by reading the help for matchMotifs, ?matchMotifs.

---
## motifmatchr inputs

motifmatchr needs some search space. Though we could run this genome-wide, for efficiency and speeds sake we normally just run it on regions of interest i.e. our W6-specific peaks. 

Specifically we need the sequence associated with these peaks. Luckily we generated all this while running MEME. 


```{r}
peaksSequences
```

---
## Finding motif positions

As we saw in its help, The matchMotifs function can provide output of motif matches as matches, scores or positions.

Here we scan for the top 3 motifs from our AME results using the sequence under the first 100 peaks and specify the out as **positions**.

```{r}
motifs_subset <- motifs[names(motifs) %in% ame_results$motif_alt_id[1:3]] 

motif_positions <- matchMotifs(motifs_subset, peaksSequences[1:100],out="positions")

class(motif_positions)
length(motif_positions)

```

---
## Finding motif positions

The result contains a list of the same length as the number of motifs tested.

Each element contains a IRangeslist with an entry for every sequence tested and a IRanges of motif positions within the peak sequences.

```{r}
motif_positions$MA1152.2
```

---
## Finding motif positions

We can unlist our IRangeslist to a standard list for easier working.

```{r}

MA1152.2hits <- motif_positions$MA1152.2
names(MA1152.2hits) <- names(peaksSequences[1:100])
unlist(MA1152.2hits, use.names = TRUE)
```
---
## Using motif positions

Once motif instances have been identified they can then be used to generate metaplots like we saw in [session 2 with soGGi](https://rockefelleruniversity.github.io/ATAC.Cut-Run.ChIP/presentations/singlepage/Session3_QC.html#Evaluating_TSS_signal), or we could use [profileplyr](https://www.bioconductor.org/packages/devel/bioc/vignettes/profileplyr/inst/doc/profileplyr.html#generate-group-annotated-heatmap-in-r-directly-with-generateenrichedheatmap). 

These metaplots would show changes to accessibility/enrichment directly over the motifs. The only changes to what we should in session 2 would be to replace the TSS ranges with motif ranges. 

---
## Finding motif hits

We may simply want to map motifs to their peaks.

To do this we can set the **out** parameter to **matches**.  This will return a SummarizedExperiment object.

```{r}
motifHits <- matchMotifs(motifs, peaksSequences, out="matches")
class(motifHits)
motifHits
```

---
## Finding motif hits

We can retrieve a matrix of matches by motif and peak using the motifMatches function.

```{r}
mmMatrix <- motifMatches(motifHits)
dim(mmMatrix)
mmMatrix[1:8,1:8]
```

---

## Finding motif hits

Although a sparse matrix, we can still use our matrix operations to extract useful information from this object.

We can use the **colSums()** to identify the total occurrence of motifs in our peak sequences.

Fittingly our top hit from AME has a high number of motifs.

```{r,eval=F, echo=T}
totalMotifOccurence <- colSums(mmMatrix)
totalMotifOccurence[1:4]
totalMotifOccurence["MA1152.2"]
```

```{r,eval=T, echo=F}
my_mat<-apply(data.matrix(mmMatrix),2,sum)
names(my_mat)<-colnames(mmMatrix)
my_mat[1:4]

my_mat["MA1152.2"]

```

---
## Finding motif hits

We can also identify peaks which contain a hit for a selected motif.


```{r,eval=T}
UpinW6_peaksWithMA1152.2 <- UpinW6_peaks[mmMatrix[,"MA1152.2"] == 1]
UpinW6_peaksWithMA1152.2
```

---
## Other motif tools

We have focussed primarily on using MEME to conduct our analysis, there are several other popular tools. 

* [*Homer*](http://homer.ucsd.edu/homer/motif/) has some motif detection tools that are conceptually similar to MEME.

* [*chromVAR*](https://github.com/GreenleafLab/chromVAR) continues on from what we have done with motifmatchr and checks for differential accessibility of peaks that do/don't contain motifs.

---
## Time for an exercise!


Exercise on functions can be found [here](https://rockefelleruniversity.github.io/ATAC.Cut-Run.ChIP/exercises/exercises/MyExercise6_exercise.html)

---
## Answers to exercise.


Answers can be found here  [here](https://rockefelleruniversity.github.io/ATAC.Cut-Run.ChIP/exercises/answers/MyExercise6_answers.html)


